core/src/test/java/org/infinispan/util/concurrent/BlockingManagerTest.java

```diff
@@ -0,0 +1,178 @@
+
+      Mockito.verify(blockingExecutor).execute(Mockito.any());
+      Mockito.verifyZeroInteractions(nonBlockingExecutor);
+   }
+
+   public void testBlockingPublishToVoidStageInvokedNonBlockingThreadCompleteAfterSubscribe() {
+      BlockingManager blockingManager = createBlockingManager(false);
+
+      AsyncProcessor<Object> processor = AsyncProcessor.create();
+
+      processor.onNext(1);
+
+      CompletionStage<Void> stage = blockingManager.blockingPublisherToVoidStage(processor
+            .doOnNext(BlockHoundHelper::blockingConsume), null);
+      assertFalse(CompletionStages.isCompletedSuccessfully(stage));
+
+      processor.onComplete();
+
+      assertTrue(CompletionStages.isCompletedSuccessfully(stage));
+
+      Mockito.verify(blockingExecutor).execute(Mockito.any());
+      Mockito.verify(nonBlockingExecutor).execute(Mockito.any());
+   }
+
+   public void testBlockingPublisherInvokedBlockingThread() {
+      BlockingManager blockingManager = createBlockingManager(true);
+
+      Publisher<Integer> publisher = blockingManager.blockingPublisher(Flowable.just(1)
+            .doOnNext(BlockHoundHelper::blockingConsume));
+
+      TestSubscriber<Integer> subscriber = TestSubscriber.create();
+      publisher.subscribe(subscriber);
+
+      subscriber.assertComplete();
+
+      // We should not have used any executor as we were a blocking thread already
+      Mockito.verifyZeroInteractions(nonBlockingExecutor, blockingExecutor);
+   }
+
+   public void testBlockingPublisherInvokedBlockingThreadCompleteAfterSubscribe() {
+      BlockingManager blockingManager = createBlockingManager(true);
+
+      AsyncProcessor<Integer> processor = AsyncProcessor.create();
+
+      processor.onNext(1);
+
+      Publisher<Integer> publisher = blockingManager.blockingPublisher(processor
+            .doOnNext(BlockHoundHelper::blockingConsume));
+
+      TestSubscriber<Integer> subscriber = TestSubscriber.create();
+      publisher.subscribe(subscriber);
+
+      subscriber.assertNotComplete();
+
+      processor.onComplete();
+
+      subscriber.assertComplete();
+
+      // We should not have used any executor as we were a blocking thread already for onNext and onComplete is done on
+      // the invoking thread as it happened after publish
+      Mockito.verifyZeroInteractions(nonBlockingExecutor, blockingExecutor);
+   }
+
+   public void testBlockingPublisherInvokedNonBlockingThread() {
+      BlockingManager blockingManager = createBlockingManager(false);
+
+      Publisher<Integer> publisher = blockingManager.blockingPublisher(Flowable.just(1)
+            .doOnNext(BlockHoundHelper::blockingConsume));
+
+      TestSubscriber<Integer> subscriber = TestSubscriber.create();
+      Flowable.fromPublisher(publisher)
+            // We should observe any value of returned Publisher from `blockingPublisher` on a non blocking thread
+            .doOnNext(ignore -> assertTrue(BlockHoundHelper.currentThreadRequiresNonBlocking()))
```

Comment:
BlockingManager thinks the subscribe thread is non-blocking and the blocking executor is actually a mock that doesn't do anything, so I'm not sure how the implementation could make this assert fail.

Reply:
Yeah, that is the point, it thinks the thread is non blocking, thus it submits the publisher to a blocking thread for the subscribe part. Thus we invoke `blockingConsume` to verify if it is in a blocking thread. Then the returned publisher has the doOnNext to verify that is done in a non blocking thread as we have to publish values on the non blocking thread. I think the confusion is the same as the prior comment above.  https://github.com/infinispan/infinispan/pull/8637#discussion_r573151661

Reply:
@wburns I was trying to say that `blockingExecutor` is just a mock, so it does not run the callback at all, therefore the `assertTrue()` call can't fail.

But I see I was wrong: `blockingExecutor` not used only as a mock, to verify which methods are being called, it also executes the task as a normal executor would.